### PR TITLE
feat(meshservice): k8s controller to convert service

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Checklist prior to review

Implementation of https://github.com/kumahq/kuma/pull/9722

- [ ] [Link to relevant issue][1] as well as docs and UI issues --
- [ ] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/runtime/k8s/controllers/meshservice_controller.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	Log    logr.Logger
+	Scheme *kube_runtime.Scheme
+}
+
+func (r *MeshServiceReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	log := r.Log.WithValues("service", req.NamespacedName)
+
+	namespace := &kube_core.Namespace{}
+	if err := r.Client.Get(ctx, kube_types.NamespacedName{Name: req.Namespace}, namespace); err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to get Namespace for Service")
```
Comment (lahabana) [inline pkg/plugins/runtime/k8s/controllers/meshservice_controller.go:45]: NotFound handling?

File: pkg/plugins/runtime/k8s/controllers/meshservice_controller.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		ms.Spec.Ports = []meshservice_api.Port{}
+		for _, port := range svc.Spec.Ports {
+			ms.Spec.Ports = append(ms.Spec.Ports, meshservice_api.Port{
+				Port:       uint32(port.Port),
+				TargetPort: uint32(port.TargetPort.IntVal), // todo(jakubdyszkiewicz): what to do about named ports?
+				Protocol:   core_mesh.Protocol(pointer.DerefOr(port.AppProtocol, "tcp")),
+			})
+		}
+
+		if svc.Spec.ClusterIP != "" {
```
Comment (lahabana) [inline pkg/plugins/runtime/k8s/controllers/meshservice_controller.go:115]: What about stateful set?

**Comment by jakubdyszkiewicz** [comment]
[] add flag to disable existing vips

File: pkg/plugins/runtime/k8s/controllers/meshservice_controller.go (hunk 2 above)
Comment (jakubdyszkiewicz) [reply pkg/plugins/runtime/k8s/controllers/meshservice_controller.go]: good point. Ignoring for now. Will add handling after Kube UX MADR is merged.

File: pkg/plugins/runtime/k8s/controllers/meshservice_controller.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+		ms.Spec.Ports = []meshservice_api.Port{}
+		for _, port := range svc.Spec.Ports {
+			ms.Spec.Ports = append(ms.Spec.Ports, meshservice_api.Port{
+				Port:       uint32(port.Port),
+				TargetPort: uint32(port.TargetPort.IntVal), // todo(jakubdyszkiewicz): update after API changes
+				Protocol:   core_mesh.Protocol(pointer.DerefOr(port.AppProtocol, "tcp")),
+			})
+		}
+
+		ms.Spec.Status.VIPs = []meshservice_api.VIP{
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/meshservice_controller.go:124]: Do we add an entry to `Status.Addresses`?

File: pkg/plugins/runtime/k8s/controllers/meshservice_controller.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+		if ms.Spec == nil {
+			ms.Spec = &meshservice_api.MeshService{}
+		}
+
+		ms.Spec.Selector = meshservice_api.Selector{
+			DataplaneTags: svc.Spec.Selector,
+		}
+
+		ms.Spec.Ports = []meshservice_api.Port{}
+		for _, port := range svc.Spec.Ports {
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/meshservice_controller.go:116]: Should we check that `port.Protocol == TCP`?

File: pkg/plugins/runtime/k8s/controllers/meshservice_controller.go (hunk 3 above)
Comment (jakubdyszkiewicz) [reply pkg/plugins/runtime/k8s/controllers/meshservice_controller.go:124]: we don't. Status addresses is to configure our DNS. We don't need to configure our DNS

File: pkg/plugins/runtime/k8s/controllers/meshservice_controller.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+func (r *MeshServiceReconciler) deleteIfExist(ctx context.Context, key kube_types.NamespacedName) error {
+	ms := &meshservice_k8s.MeshService{}
+	if err := r.Client.Get(ctx, key, ms); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrap(err, "could not get mesh service")
+	}
+	if err := r.Client.Delete(ctx, ms); err != nil {
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/meshservice_controller.go:157]: alternatively just delete it and ignore the not found error

File: pkg/plugins/runtime/k8s/controllers/meshservice_controller.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+func MeshToMeshService(l logr.Logger, client kube_client.Client) kube_handler.MapFunc {
+	l = l.WithName("mesh-to-service-mapper")
+	return func(ctx context.Context, obj kube_client.Object) []kube_reconcile.Request {
+		services := &kube_core.ServiceList{}
+		if err := client.List(ctx, services); err != nil {
+			l.WithValues("namespace", obj.GetName()).Error(err, "failed to fetch Services")
+			return nil
+		}
+		var req []kube_reconcile.Request
+		for _, pod := range services.Items {
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/meshservice_controller.go:200]:
```suggestion
		for _, svc := range services.Items {
```

**Review by michaelbeaumont — APPROVED** [review]

**Pull request merged**